kratos/solving_strategies/builder_and_solvers/p_multigrid/master_slave_constraint_assembler.cpp

```diff
@@ -0,0 +1,345 @@
+struct MasterSlaveConstraintAssembler<TSparse,TDense>::Impl
+{
+    std::vector<std::size_t> mSlaveIds;
+
+    std::vector<std::size_t> mMasterIds;
+
+    std::unordered_set<std::size_t> mInactiveSlaveIds;
+
+    std::unique_ptr<Scaling> mpDiagonalScaling;
+
+    int mVerbosity;
+};
+
+
+template <class TSparse, class TDense>
+MasterSlaveConstraintAssembler<TSparse,TDense>::MasterSlaveConstraintAssembler() noexcept
+    : MasterSlaveConstraintAssembler(Parameters())
+{}
+
+
+template <class TSparse, class TDense>
+MasterSlaveConstraintAssembler<TSparse,TDense>::MasterSlaveConstraintAssembler(Parameters Settings)
+    : MasterSlaveConstraintAssembler(Settings, "unnamed")
+{}
+
+
+template <class TSparse, class TDense>
+MasterSlaveConstraintAssembler<TSparse,TDense>::MasterSlaveConstraintAssembler(Parameters Settings,
+                                                                               std::string&& rInstanceName)
+    : Base(ConstraintImposition::MasterSlave, std::move(rInstanceName)),
+      mpImpl(new Impl)
+{
+    KRATOS_TRY
+    // Parse diagonal scaling and validate other settings.
+    Parameters default_parameters = this->GetDefaultParameters();
+    Parameters default_diagonal_scaling = default_parameters["diagonal_scaling"].Clone();
+    default_parameters.RemoveValue("diagonal_scaling");
+    std::optional<Parameters> maybe_diagonal_scaling = Settings.Has("diagonal_scaling") ? Settings["diagonal_scaling"].Clone() : std::optional<Parameters>();
+    if (maybe_diagonal_scaling.has_value()) Settings.RemoveValue("diagonal_scaling");
+    Settings.ValidateAndAssignDefaults(default_parameters);
+    Settings.AddValue("diagonal_scaling", maybe_diagonal_scaling.has_value() ? maybe_diagonal_scaling.value() : default_diagonal_scaling);
+    mpImpl->mpDiagonalScaling = std::make_unique<Scaling>(Settings["diagonal_scaling"]);
+    mpImpl->mVerbosity = Settings["verbosity"].Get<int>();
+    KRATOS_CATCH("")
+}
+
+
+template <class TSparse, class TDense>
+MasterSlaveConstraintAssembler<TSparse,TDense>::~MasterSlaveConstraintAssembler() = default;
+
+
+template <class TSparse, class TDense>
+void MasterSlaveConstraintAssembler<TSparse,TDense>::Allocate(const typename Base::ConstraintArray& rConstraints,
+                                                              const ProcessInfo& rProcessInfo,
+                                                              typename TSparse::MatrixType& rLhs,
+                                                              typename TSparse::VectorType& rSolution,
+                                                              typename TSparse::VectorType& rRhs,
+                                                              typename Base::DofSet& rDofSet)
+{
+    KRATOS_TRY
+    std::vector<std::unordered_set<IndexType>> indices(rDofSet.size());
+    std::vector<LockObject> mutexes(indices.size());
+    const auto it_const_begin = rConstraints.begin();
+
+    #pragma omp parallel
```

Comment:
Not utils?

Reply:
This is a copy-pasta from `ResidualbasedBlockBuilderAndSolver`. The reason it can't use parallel utils is because the for loops are `nowait`.